File: credit.c
#include <cs50.h>
#include <stdio.h>
#include <ctype.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

bool check_number(string number);

int main(void)
{
    string number;
    do
    {
        number = get_string("Number:");

    }
    while(!check_number(number));

        int N = strlen(number);
        int sum_2 = 0;
        int sum = 0;
        char n_1;
        char n_2;
        int num;
        int num_2;
        if (N % 2 > 0)
        {
            for (int i = 0 ; i < N ; i += 2 )
            {
                n_1 = number[i];
                num = number[i] - '0';
                if (i == 0)
                {
                    sum += num;
                }
                else{

                num_2 = number[i - 1] - '0';
                if ((2*num_2) % 10 > 0)
                {
                    num_2 = 2*num_2;
                    int num_new_2 = (num_2 % 10) + (num_2 / 10);
                    sum_2 += num_new_2;
                }
                else{

                sum_2 += 2* (num_2);


                }

                sum += num;

                }
            }

        }
        if (N % 2 == 0)
        {
            for (int i = 0 ; i < N ; i += 2 )
            {
                num_2 = number[i] - '0';

                if ((2*num_2) >= 10)
                {
                    num_2 = 2*num_2;
                    int num_new_2 = (num_2 % 10) + (num_2 / 10);
                    sum_2 += num_new_2;
                }
                else{

                sum_2 += 2* (num_2);

                }

                num = number[i + 1] - '0';
                sum_2 += num;

                }

            }



        if ((sum + sum_2) % 10 == 0 )
        {
            int first_digit = number[0] - '0';
            int sec_digit = number[1] - '0';

            switch(first_digit){
                case 4:
                if (strlen(number) == 13 || strlen(number) == 16)
                {
                printf("VISA\n");
                }
                else
                {
                    printf("INVALID\n");
                }
                break;
                case 5:
                if ((sec_digit == 1 || sec_digit == 2 || sec_digit == 3 || sec_digit == 4 || sec_digit == 5) && strlen(number) == 16)
                {
                    printf("MASTERCARD\n");
                }
                else{
                printf("INVALID\n");
                }
                break;
                case 3:
                if ((sec_digit == 4 || sec_digit == 7) && strlen(number) == 15)
                {
                    printf("AMEX\n");
                    break;
                }
                else{
                printf("INVALID\n");
                break;
                }
                default:
                printf("INVALID\n");
                break;
                }
        }
        else
        {
        printf("INVALID\n");
        }

        return 0;
    }

bool check_number(string number)
{
    int digit;
    int length = strlen(number);
    for (int i = 0; i < length; i++)
    {
        if (!isdigit(number[i]))
        {
            return false;
            break;
        }
    }
    return true;
}

File: credit.py
# TODO
# Going to pass the number into this list so easier to apply Luhn's algorithm
digits = []
# Function to store the number into the list
def store_digits(num):
    if num == 0:
        return
    else:
        last_digit = int(num % 10)
        digits.append(last_digit)
        store_digits(int(num / 10) )

# Prompt the user for a number
# If wrong input, infinite prompt until user provides proper input
while True:
    try:
        print("Number:")
        number = int(input())
        store_digits(number)
        break
    except:
        print("Invalid number")

length = len(digits) # length of the number
sum = 0 # To keep track of the sum of the digits not multiplied by 2
sum_2 = 0 # To keep track of the sum of the digits multiplied by 2

if int(len(digits) % 2) == 0:

    for i in range(0, length, 2):
        num_2 = digits[i + 1]
        if 2*num_2 >= 10:
            num_2 = int(2*num_2)
            num_new_2 = int(num_2 % 10) + int(num_2 / 10)
            sum_2 += num_new_2
        else:
            sum_2 += 2*num_2
        num = digits[i]
        sum += num


if int(len(digits) % 2 ) > 0: ## **fix**

    for i in range(0, length, 2):
        num = digits[i]
        if i == 0:
            sum += num
        else:
            num_2 = digits[i - 1]
            if 2*num_2 >= 10:
                num_2 = 2*num_2
                num_2 = int(num_2 % 10) + int(num_2 / 10)
                sum_2 += num_2
            else:
                sum_2 += 2*num_2
            sum += num

if ((sum + sum_2) % 10) == 0:
    sec_digits = [1, 2, 3, 4, 5]
    if digits[length - 1] == 4 and (len(digits) == 13 or len(digits) == 16):
        print("VISA")
    else:
        print("INVALID")
    if digits[length - 1] == 5:
        for i in range(5):
            if digits[length - 2] == sec_digits[i]:
                print("MASTERCARD")
            else:
                print("INVALID")

    if digits[length - 1] == 3:
        if digits[length - 2] == 4 or digits[length - 2] == 7:
            print("AMEX")
        else:
            print("INVALID")
else:
    print("INVALID")

print()
